### ISSUE TITLE
Offer hooks or plugin interface for non git deployment

### ISSUE THREAD

**Issue by couling** [jimporter/mike#131]
I've come here from a very similar angle to #31 . Mike is recommended by material. The feature of having every released version of the documentation available became our primary reason for adopting mike.

Yet working with deployment direct to git doesn't work so well in some contexts outside GitHub pages.

Besides the point raised in #31 theres some real complexity in keeping the actual deployment in sync with the git branch.  AFAIK there's no easy way to apply a git patch to an S3 bucket (static website).  A full sync breaks the principle of leaving deployed versions untouched and has gotyas that mean it's not totally reliable.

Honestly I'm wondering if Mike is really the right tool for us.  But I keep coming back to the fact it it has a really neatly packaged mechanism for writing alias forwarding html files and versions.json.  It offers a slick command line interface for managing deployed versions and aliases.

Nobody wants to reinvent the wheel.

--------

I wonder if you would consider opening up some kind of plugin interface or hooks that could be used to deploy somewhere other than a git branch.

I realise Mike was imagined as a git only tool.

-----

Functionally Mike fetches some files from a git commit, invokes mkdocs, generates some simple files, and commits back. (Maybe I missed something)

The kind of hook I'm imagining is one which would allow a plugin to swap in different behaviour for fetching and writing / pushing files.

A plugin developer could then implement, for example, an S3 client leveraging Mike to maintain files in an S3 bucket instead of files in a git branch

**Comment by couling** [comment]
Okay I respect that, thanks for your time.

The reason for not rsync and similar is that Mike knows which files have changed, it writes them. Other things have to scan and compare. That's slow and there are reasons it can fail to sync correctly as it's usually based on size and file time.

Ultimately I was trying to doge writing my own from scratch.  But I guess it's not such a huge piece of code.

**Comment by jimporter** [comment]
If you do write something like that (and open source it), I'd be happy to provide a link to it from mike. mike was never intended to be the universal solution for this sort of thing, but a set of broadly-compatible tools that work in different situations would certainly be good for the "mike ecosystem".

**Comment by jimporter** [comment]
> The reason for not rsync and similar is that Mike knows which files have changed, it writes them. Other things have to scan and compare. That's slow and there are reasons it can fail to sync correctly as it's usually based on size and file time.

You could get the list of changed files from Git: `git diff-tree --no-commit-id --name-only -r HEAD^`. Then it's just a matter of feeding those file names into your deploy script.

**Comment by couling** [comment]
> You could get the list of changed files from Git: git diff-tree --no-commit-id --name-only -r HEAD^

It's more like this: assuming we store the last deploy commit id somewhere and know it is`b7be3afad600530196e3d90dc82733b7e5ad8693` then it would be something more like, find the `gh-pages` commit id with `git show-ref refs/heads/gh-pages` then

```shell
# Don't checkout, that can mess up the CI tool
rm -rf staging_directory
git clone --branch afd883f4988546967a064cfb0e31fb60d6ba3a0f . staging_directory

git log --pretty=oneline b7be3afad600530196e3d90dc82733b7e5ad8693..afd883f4988546967a064cfb0e31fb60d6ba3a0f | while read commit other
do
  git diff-tree --no-commit-id --name-only -r $commit
done | sort -u | while read file
do
  if [ -f "$file" ]
  then
    # Push "staging_directory/$file" to remote_url/$file
  else
    # Delete remote_url/$file
  fi
done
# Push afd883f4988546967a064cfb0e31fb60d6ba3a0f to a file so we know what we updated to.
```

For targets other than S3 this would also involve a sanity check of replacing a directory with a file or visa versa and the obvious need to make directories on the remote.  

I'm not sure I've caught every edge case.

**Comment by jimporter** [comment]
Yeah, from a certain point of view, you could look at this problem as follows: mike puts the authoritative state of your documentation site in `gh-pages`, and then you just need to figure out the best way to sync those changes with the server. Since Git has lots of plumbing to let you inspect changes, this shouldn't be too hard.

However, this strategy would mean that you're not getting any of the extra niceties (e.g. fancy redirection rules) that S3 static sites could provide.
